### PR TITLE
perf(io): lazy regional crop via tmd.crop

### LINKED ISSUE #612: FES netCDF open loads full global grids before crop (slow regional extract)
## Summary

Opening FES2014/FES2022 constituent netCDFs via `open_fes_netcdf` / `open_mfdataset` / `model.open_dataset` reads **entire global** amplitude/phase fields for every constituent, even when the caller only needs a small AOI (e.g. estuary open-boundary nodes).

The documented [Cropping](https://pytmd.readthedocs.io/en/stable/user_guide/Cropping.html) path (`ds.tmd.crop`) runs **after** that load. With dask chunks present, `crop` currently does `chunk(-1).compute()` before `where`, so [Chunking](https://pytmd.readthedocs.io/en/stable/user_guide/Chunking.html) alone does not avoid the global I/O for multi-file FES.

Regional SCHISM/rompy tidal BC setups (dozens of FES2022 files × global 1/16° grids) become impractical.

## Environment

- pyTMD 3.0.9
- FES-netcdf (FES2014 / FES2022)

## Expected

Optional `bounds` on FES netCDF open so each file hyperslabs lon/lat **before** building the complex constituent field (same idea as pre-xarray `read_netcdf_file(..., bounds=)`). Callers expand the box themselves if they need a halo for interpolation.

```python
ds = m.open_dataset(
    group="z",
    constituents=["m2", "s2"],
    bounds=[149.5, 152.5, -24.5, -21.5],  # xmin, xmax, ymin, ymax (caller-buffered)
)
```

## Proposed fix (initial)

1. **`open_fes_netcdf`**: accept `bounds`; index-slice amp/phase (and coords) before `amp * exp(-i*pha)`.
2. Thread kwargs through existing `open_mfdataset` → `model.open_dataset` (already forwards `**kwargs`).
3. Follow-up (separate PR): make `Dataset.tmd.crop` use index/`sel` without full `.compute()` so post-open crop stays lazy.

## Related

- Cropping / Chunking recipes already recommend regional workflows; this makes open-path match that intent.
- [#537](https://github.com/pyTMD/pyTMD/issues/537) (extrapolate fill values) is orthogonal correctness; regional open still helps those workflows.

## Benchmark (local, include in issue/PR)

Script (not pytest-collected): `test/perf_fes_windowed_read.py` on branch
`perf/fes-windowed-netcdf-read`.

```bash
PYTHONPATH=. python test/perf_fes_windowed_read.py \
  --tide-dir $TIDE_MODEL_PATH \
  --repeats 3
```

**Machine run (2026-08-04):** FES2022 ocean_tide netCDF
`[149.5, 152.5, -24.5, -21.5]`, 3 repeats, median.

| Case | Full open | `bounds=` open | Notes |
|------|-----------|----------------|-------|
| single file (`m2_fes2022.nc`) | 1.221 s | 0.009 s | **130×**; shape (5401, 10801)→(91, 91); 466.7→0.07 MB |
| `open_mfdataset` (8 constituents) | 10.040 s | 0.083 s | **121×**; m2 s2 k1 o1 n2 k2 p1 q1 |

Full 34-constituent FES2022 open would scale roughly linearly with file count
on the full path; windowed open stays near-constant per file at the AOI size.

## Local branch

`perf/fes-windowed-netcdf-read` on https://github.com/pbranson/pyTMD

### PR DESCRIPTION
**Pull request by pbranson**
## Summary
- Add `isel_bounds` and make gridded `Dataset.tmd.crop` / `DataTree.tmd.crop` use index hyperslabs (lazy for chunked/netCDF), with dateline-safe Pacific longitude windows (no half-globe pad).
- Treat FES-style `0…360` inclusive lon grids as global so wrap crop engages.
- Prefer chunked open then `tmd.crop` for regional FES reads; document that path. Add focused tests and an optional benchmark script (`--tide-dir` / `TIDE_MODEL_PATH`).

## Benchmark (local, FES2022)
| Case | Full open | Chunked open + crop |
|------|-----------|---------------------|
| Fitzroy box single file | ~1.25 s | ~0.026 s (~48×) |
| Dateline `[170,-170]` single file | ~1.25 s | ~0.038 s (~33×) |

## Test plan
- [x] `pytest test/test_fes_windowed_read.py`
- [x] `pytest test/test_fes_predict.py` (local, when tide cache present)
- [x] flake8 syntax select (CI-equivalent) on `pyTMD` + `test`
- [x] `ruff format --check` on touched files
- [x] CI on this PR

### CODE REVIEW

**Comment by tsutterley** [comment]
@pbranson this is super cool and provides a huge speed up! My main comment is to maybe move `_isel_bounds` to `io.dataset` and import it in the individual readers. Seems like other the other read programs (e.g. `ATLAS-netcdf`) would benefit too.

**Comment by pbranson** [comment]
Thats a good suggestion, I thought it would be more widely relevant but didnt want to over step the mark. 

I reviewed the io.datasets and realised that you are using datatree there so I made the changes to be a bit more native xarray/datatree - the hyperslab works when opening with chunking so the indexing is performed on lazy datasets prior to large IO. This more general approach doesnt have the same speedup as the lower level change for jsut FES due to the broader call stack, but this addresses all model drivers implicitly

I also added some more tests for dateline crossing crops to make sure it is robust for that edge case too. 

Apologies about the noise on the PR branch, the AI coding agent got a little over zealous!

**Comment by tsutterley** [comment]
@pbranson tag me when you are ready for a review :smile:

**Comment by pbranson** [comment]
@tsutterley Yes please review when you have time. Thanks!

File: pyTMD/io/FES.py [hunk 1]
```diff
@@ -338,7 +345,7 @@ def open_fes_netcdf(
     # amplitude and phase variable names
     amp_key = mapping_amp[kwargs["group"]]
     phase_key = mapping_ph[kwargs["group"]]
-    # mask where amplitude or phase are zero
+    # mask where amplitude and phase are zero
```
[the remark this answers is not in the document]
Comment (tsutterley) [reply pyTMD/io/FES.py]: @pbranson this is on me and your description is right. Phase can be 0 and still have a valid amplitude. I think this should be
```python
valid_values = (tmp[amp_key] != 0) | (tmp[phase_key] != 0)
```

File: pyTMD/io/dataset.py [hunk 2]
```diff
@@ -1393,6 +1398,192 @@ def _has_compatible_units(self):
...
+            # place the first segment near the requested start longitude
+            while (np.min(x) - xmin) > 180.0:
+                x -= 360.0
+            while (xmin - np.min(x)) > 180.0:
+                x += 360.0
+        else:
+            # keep subsequent segments east of the previous piece
+            while np.min(x) <= x_cursor + 1e-12:
+                x += 360.0
+        pieces.append(part.assign_coords(x=("x", x)))
```
[the remark this answers is not in the document]
Comment (pbranson) [reply pyTMD/io/dataset.py:1544]: Good catch. The wrap tests used 0…359, so they missed FES-style duplicated 0/360. On those grids the west piece kept 360 and the east piece’s 0 then got an extra +360 shift (e.g. [350, 370] became [350…360, 720…730]). Dropped the closing meridian from the first wrap piece in b94874c; tests now cover inclusive 0…360 and −180…180.

File: pyTMD/io/FES.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pbranson) [reply pyTMD/io/FES.py]: OK cool, have switched to or and updated the comment accordingly.

**Comment by tsutterley** [comment]
thanks @pbranson! we're merged :)
